### PR TITLE
server: balance by namespace.

### PR DESCRIPTION
**Pull request by disksing**
refer #742

### CODE REVIEW

File: server/namespace_cluster.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	for _, s := range allStats {
+		if c.GetRegion(s.RegionID) != nil {
+			stats = append(stats, s)
+		}
+	}
+	return stats
+}
+
+func scheduleByNamespace(cluster schedule.Cluster, classifier namespace.Classifier, scheduler schedule.Scheduler) *schedule.Operator {
+	for _, ns := range classifier.GetAllNamespaces() {
```
Comment (dantin) [inline server/namespace_cluster.go]: Is there any chance that the region on the beginning of the the namespace list got scheduled, while the ones on the end of the list not

File: server/namespace/classifer.go [hunk 2]
```diff
@@ -19,19 +19,24 @@ import "github.com/pingcap/pd/server/core"
...
 
 // Classifier is used to determine the namespace which the store or region
 // belongs.
 type Classifier interface {
+	GetAllNamespaces() []string
 	GetStoreNamespace(*core.StoreInfo) string
 	GetRegionNamespace(*core.RegionInfo) string
 }
 
 type defaultClassifer struct{}
```
[the remark this answers is not in the document]
Comment (disksing) [reply server/namespace/classifer.go]: Sure.

File: server/namespace_cluster.go (hunk 1 above)
Comment (disksing) [reply server/namespace_cluster.go]: Yes, it's possible. I'll change it to pick namespace in random order.

Comment (dantin) [reply server/namespace_cluster.go]: LGTM

File: server/namespace/classifer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dantin) [reply server/namespace/classifer.go]: LGTM

**Comment by dantin** [comment]
LGTM

File: server/namespace/classifer.go [hunk 3]
```diff
@@ -19,23 +19,28 @@ import "github.com/pingcap/pd/server/core"
 // default.
 const DefaultNamespace = "global"
 
-// DefaultClassifer is a classifier that classifies all regions and stores to
+// DefaultClassifier is a classifier that classifies all regions and stores to
 // DefaultNamespace.
-var DefaultClassifer = defaultClassifer{}
+var DefaultClassifier = defaultClassifier{}
```
[the remark this answers is not in the document]
Comment (siddontang) [reply server/namespace/classifer.go]: I think we can use a register mechanism.

[the remark this answers is not in the document]
Comment (disksing) [reply server/namespace/classifer.go]: The `GetClassifier` function is supposed to be in package `server`, maybe as a method of `RaftCluster`. The `RaftCluster` can create a classifier based on config, then use it to create a `coordinator` instance.

[the remark this answers is not in the document]
Comment (ming-relax) [reply server/namespace/classifer.go]: @disksing ok

**Comment by ming-relax** [comment]
LGTM

File: server/namespace_cluster.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	}
+	return r
+}
+
+// RegionWriteStats returns hot region's write stats.
+func (c *namespaceCluster) RegionWriteStats() []*core.RegionStat {
+	allStats := c.Cluster.RegionWriteStats()
+	stats := make([]*core.RegionStat, 0, len(allStats))
+	for _, s := range allStats {
+		if c.GetRegion(s.RegionID) != nil {
```
Comment (nolouch) [inline server/namespace_cluster.go]: should also check region

Comment (disksing) [reply server/namespace_cluster.go]: `GetRegion` checks namespace inside.

Comment (nolouch) [reply server/namespace_cluster.go]: got it

**Pull request merged**